### ISSUE TITLE
Some Questions about Code Implements.

### ISSUE THREAD

**Issue by xiaobanni** [watchernyu/REDQ#8]
Thanks for this excellent work!
I have some questions about the code implements.

1.  In `core.py`  line 214, you do `torch.clamp()` to `log_std`.  Why we need `clamp()` here, could it be that `log_std` will diverge if not `clamp`?  And if the `log_std` is truncated, how should the gradient propagate (This seems to be a question about torch, but I didn't find a good answer)?
2.  In `redq_sac` function `get_redq_q_target_no_grad`, you use `policy.forward(obs_next_tensor)` with default parameters `deterministic==False`, why we don't use `deterministic==True`? The above confusion may also be paraphrased as, the `Critic network`, _i.e._ `Q_function` here, is approximating the policy state-action value function $Q_\pi$, or is approximating the optimal state-action value function $Q_*$?

Looking forward to your answer and thanks in advance.

**Comment by watchernyu** [comment]
Thanks for posting the issue! These are good questions!

1. since we mainly tested REDQ build on top of SAC, for this one we basically used what the SAC authors did. In the original SAC codebase, there is this truncation. I don't recall if the authors mentioned this detail in their paper (perhaps not). So they put this here perhaps just to make sure the value is not crazy. Now, if it's truncated, then there will be no gradient because the clamp or truncation operation is clearly not differentiable (there is another type of clipping where you will rescale it to a smaller value, in that case there will be gradient).

2. Very good question, this is again, a detail that comes from the SAC/TD3 codebase. (I think it might be mentioned first in the TD3 paper and some concurrent work), you can check here https://spinningup.openai.com/en/latest/algorithms/sac.html#id1 and search for the word `smoothing` on this page. And also for the TD3 paper "Addressing Function Approximation Error in Actor-Critic Methods" you will find some discussion. Essentially these authors argue that this is to make your policy "smoother", and a number of papers have found that this will help make RL works better. 

So this issue might be slightly different from your question that whether it's approximating the optimal Q function. For the Q function, I think it's always trying to approximate the optimal Q, since SAC is fundamentally a Q-learning method (but in continuous action space), you can check the Sutton and Barto book and look at the Q learning section. 

Since some of these details are not invented by us, I'm not 100% confident in my answers, but I hope they are helpful!  Let us know if you have other questions!

**Comment by watchernyu** [comment]
Just to add a short comment, since in the paper we are trying to isolate the effect of bias control under high update ratio, we tried to make everything the same as SAC as much as possible without adding in small hacks. So majority of this codebase is just SAC. You can check the openai spinup page for some nice details of SAC (and also TD3, since SAC also used some important tricks in TD3), the spinup is a very good resource. 
https://spinningup.openai.com/en/latest/algorithms/sac.html#soft-actor-critic

**Comment by xiaobanni** [comment]
Thanks very much for your answer, it helped me a lot. I will look carefully at the material you gave.

**Comment by watchernyu** [comment]
Closing the issue since it is solved.
